nixos/caddy: add options for ports and openFirewall

mirkolenz opened the pull request:
## Description of changes

Added the option `services.caddy.openFirewall`. Fixes #338015

Maybe it would be a good idea to trigger a backport for 24.05 as well?

I ran the test `nix-build -A nixosTests.caddy` in a virtual NixOS machine on my Mac and got the following error. My assumption is that this is caused by some networking issues due to the nested VMs.

```txt
error: builder for '/nix/store/1b7bgcx2xf84i3aq8abgbbrrn3mwpmsd-vm-test-run-caddy.drv' failed with exit code 1;
       last 25 log lines:
       > webserver: waiting for TCP port 80 on localhost
       > webserver # Connection to localhost (::1) 80 port [tcp/http] succeeded!
       > (finished: waiting for TCP port 80 on localhost, in 1.91 seconds)
       > subtest: config is reloaded on nixos-rebuild switch
       > webserver: must succeed: /nix/store/83r48jqiag13psakxg142mnyr8jvm9mx-nixos-system-webserver-test/specialisation/config-reload/bin/switch-to-configuration test >&2
       > webserver # [  266.468913] nixos[1038]: switching to system configuration /nix/store/k19q5gmd8f0kxdavbmsvqkw4wahbgw2j-nixos-system-webserver-test
       > webserver # [  266.556710] systemd[1]: Stopped target Local File Systems.
       > webserver # [  266.659840] systemd[1]: Stopped target Remote File Systems.
       > webserver # activating the configuration...
       > webserver # setting up /etc...
       > webserver # [  295.540848] systemd[1]: Reload requested from client PID 1038 ('.switch-to-conf') (unit backdoor.service)...
       > webserver # [  295.588781] systemd[1]: Reloading...
       > webserver # restarting sysinit-reactivation.target
       > webserver # Failed to restart sysinit-reactivation.target: Did not receive a reply. Possible causes include: the remote application did not send a reply, the message bus security policy blocked the reply, the reply timeout expired, or the network connection was broken.
       > webserver # Error: Failed to get unit caddy.service
       > webserver # 
       > webserver # Caused by:
       > webserver #     Did not receive a reply. Possible causes include: the remote application did not send a reply, the message bus security policy blocked the reply, the reply timeout expired, or the network connection was broken.
       > webserver: output:
       > Test "config is reloaded on nixos-rebuild switch" failed with error: "command `/nix/store/83r48jqiag13psakxg142mnyr8jvm9mx-nixos-system-webserver-test/specialisation/config-reload/bin/switch-to-configuration test >&2` failed (exit code 1)"
       > cleanup
       > kill machine (pid 10)
       > qemu-kvm: terminating on signal 15 from pid 7 (/nix/store/h3i0acpmr8mrjx07519xxmidv8mpax4y-python3-3.12.5/bin/python3.12)
       > (finished: cleanup, in 0.01 seconds)
       > kill vlan (pid 8)
       For full logs, run 'nix log /nix/store/1b7bgcx2xf84i3aq8abgbbrrn3mwpmsd-vm-test-run-caddy.drv'.
```

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

mirkolenz committed:
nixos/caddy: add port options

drupol reviewed (commented):
I like the idea, but I would not assume that it runs by default on 80/443.

Also, what if I just want to open 443 and not 80? This option is then useless.

mirkolenz committed:
nixos/caddy: add openFirewall option

mirkolenz commented:
Thanks for having a look @drupol

I agree that my solution was not ideal for people that used custom http/https ports. I just pushed a new revision of my change and added two new options for setting the ports used by caddy. I split them up in separate commits for better traceability.

W.r.t. the option of disabling http traffic entirely: caddy by default uses the http port to forward to https (https://caddyserver.com/docs/automatic-https#effects). There also does not exist an option to disable the http port globally, one must disable automatic redirects to achieve this. So I think it should be safe to set the port globally.

However, this could be considered a breaking change: What if someone already extended the global settings of caddy with custom ports. By using types.lines, they would be merged, but I am not sure how caddy handles duplicate settings.

Therefore, I marked this PR as draft to prevent accidental merges before we have some clarity on this point.

drupol reviewed (changes requested):
This is also going to open 2 ports.

mirkolenz commented:
One way of dealing with the breaking change could be to conditionally add the port settings only if they differ from the default (just like email is only added if specified by the user).

mirkolenz commented:
@drupol I changed the default value of the ports to null and conditionally open all ports on the firewall that are set to some value. What do you think?

drupol reviewed (approved):
Diff LGTM, would you mind adding an entry in the release note ?

mirkolenz commented:
Sure, I'll have a look later today and will mark the PR as ready for review once I'm done. Thanks!

mirkolenz committed:
nixos/doc: add release notes about caddy ports/firewall

drupol reviewed (approved):
LGTM, let's wait for another review before merging just to make sure.

emilylange reviewed (commented):
Hard no from me, sorry.

Our other webserver modules like `nginx` and `apache-httpd` don't have an `openFirewall` option either.

I get why this has been structured and implemented as is.
I just don't think this module needs even more options that simply repeat Caddyfile global options.  

And finally, this is missing `udp/443` for `http3` which is enabled by default.

Maybe the other `pkgs.caddy` maintainers see more need for this than me: @Br1ght0ne @techknowlogick

mirkolenz commented:
I just thought about the port options again and am not sure if `null` should be the default value. I had a look at some other NixOS modules that have a port option and the standard seems to be to use the real port number, so users could be confused by my solution.

Would it be better to have 80/443 as default with the option of setting the to null to remove them from the global config and the firewall? I personally think so, but as said earlier this could be a breaking change for some.

What do you think @drupol?

mirkolenz commented:
I totally see your point @emilylange, feel free to close the PR if the others agree.

In case it is still wanted I could add the https port ot the list of open udp ports, thanks for pointing that out.

drupol commented:
I let @emilylange decide on this one ;)

techknowlogick commented:
Thanks for the ping @emilylange.

This PR is definitely well done, and I can see the use-case for it. That being said, I'm not the target audience for it as I prefer to work directly with fw rules, especially due to specific interfaces.

Some feedback I'd give is similar to @emilylange's in that http/3 requires udp open on 443. So instead of having two options being http and https, then perhaps have the two options be tcpports and udpports. Especially since caddy can have extensions that handle more than just http/s.

I'm not sure that this should be blocked because the pattern doesn't exist in apache-httpd or nginx, but rather I think if a pattern is introduced then we should consider what the implications are for the others as we don't want to create something that works for caddy but not the others.

While I wouldn't use these settings, I do understand that there are multiple valid ways to manage systems and "my" way isn't the only way (or even the "best" way).

I'd prefer if this became more generic, and that perhaps nginx & httpd module maintainers were looped in to see if this could work for them too, but I won't block on it.

Apologies everyone for the novel, but hopefully it's been helpful and not at all yakshaving.

mirkolenz commented:
Thank you for the input, @techknowlogick!

Based on the recent feedback, I made some changes to my PR as a suggestion for further discussion on how to handle this best. The following changes were made:

- The ports are now set to their respective values (i.e., 80/443) by default, but still allow being set to null.
- The global port options are only added to the generated Caddyfile if they are non-null and differ from their respective default values (to handle the potential breaking change of existing configs)
- The UDP port is now opened as well if openFirewall is set to true
- An additional node has been added to openFirewall saying that non-standard ports need to be opened manually

This may not be the ideal solution for all use cases, but when looking at existing NixOS modules (e.g., samba), it seems that opening the standard/default ports for a service is also done elsewhere.

I also thought about the suggestion of allowing users to specify the desired tcp/udp ports manually, but that would mean that we (i) duplicate the firewall module and (ii) require them to change the ports in two places (in the Caddyfile and in the firewall). If desired I could add another option (e.g., `extraTCPPorts` and `extraUDPPorts`), but I would argue that these more advanced use cases are better fitted by setting the options manually at the right places.

Again, I very much understand if this causes more problems than solutions, I just wanted to propose another alternative 😄

fin444 commented:
Also worth noting Caddy allows using any number of arbitrary ports for both HTTP and HTTPS.

dweee commented:
@fin444 which other webservers also allow for and has already been mentioned in an earlier comment? you have just said you would like to remind everyone of that fact? although since it is already out of the norm to not use 80/443 for webservers, i would be willing to bet that anyone doing so would figure out that they would need to open up those ports on the firewall...

mirkolenz commented:
Hi there! Any updates on this PR? If the feature is not desired due to the reasons discussed here I'm happy to close it 😄 Otherwise I could rebase it on the latest master and resolve the merge conflicts.

mirkolenz commented:
No worries, I can rebase and re-open it. Just wanted to do some clean-up of stale PR's for the new year. I'll try to push something next week.
